### PR TITLE
Feat(virtual-outbound): Add entity for virtual-outbound

### PR DESCRIPTION
**Pull request by lahabana**
### Summary

Virtual-outbounds enable users to define arbitrary hostnames.

It enables features like:

- Pod routing for stateful services
- Transition to Kuma without changing host names.

### Documentation

- [x] Link to the website [documentation PR](https://github.com/kumahq/kuma-website/pull/508)

### Testing

- [x] Unit tests
- [x] E2E tests
- [x] Manual testing on Universal
- [x] Manual testing on Kubernetes 

### Backwards compatibility

- [ ] Add `backport-to-stable` label if the code is backwards compatible. Otherwise, list breaking changes.

This relies on changes not backported already.

### CODE REVIEW

**Comment by lahabana** [comment]
Do you want me to merge this without the implem behind it to make things more reviewable?

This only registers the new policy everywhere I'll then be able to connect it and add evaluation in a follow up PR.

File: api/mesh/v1alpha1/virtual_outbound.proto [hunk 1]
```diff
@@ -0,0 +1,51 @@
...
+    message TemplateParameter {
+      // Name the name of the template parameter (must be alphanumeric).
+      string name = 1;
+      // TagKey the name of the tag in the Kuma outbound (optional if absent it
+      // will use Name).
+      string tag_key = 2;
+    }
+  }
+
+  Conf conf = 3;
```
Comment (bartsmykla) [inline api/mesh/v1alpha1/virtual_outbound.proto:50]: could you add doc annotations for fields which are required, similar to: https://github.com/kumahq/kuma/pull/2556/files?

**Review by bartsmykla — APPROVED** [review]
lgtm

File: pkg/dns/virtual_outbound_selector.go [hunk 2]
```diff
@@ -0,0 +1,57 @@
...
+
+type voMatch struct {
+	rank         mesh_proto.TagSelectorRank
+	bestSelector mesh_proto.TagSelector
+	vob          *core_mesh.VirtualOutboundResource
+}
+
+func Match(vos []*core_mesh.VirtualOutboundResource, tags map[string]string) []*core_mesh.VirtualOutboundResource {
+	var matchingVirtualOutbounds []voMatch
+	byNames := map[string]bool{}
```
Comment (lobkovilya) [inline pkg/dns/virtual_outbound_selector.go:18]: why do we need `byNames`, is it possible to have duplicates in `vos`?

File: pkg/core/resources/apis/mesh/virtual_outbound_validator.go [hunk 3]
```diff
@@ -0,0 +1,104 @@
...
+	err.Add(t.ValidateConf())
+	return err.OrNil()
+}
+
+func (t *VirtualOutboundResource) validateSelectors() validators.ValidationError {
+	return ValidateSelectors(validators.RootedAt("selectors"), t.Spec.GetSelectors(), ValidateSelectorsOpts{
+		RequireAtLeastOneSelector: true,
+		ValidateSelectorOpts: ValidateSelectorOpts{
+			RequireAtLeastOneTag: true,
+			RequireService:       true,
```
Comment (lobkovilya) [inline pkg/core/resources/apis/mesh/virtual_outbound_validator.go:23]: Why do we require a service tag?

File: api/mesh/v1alpha1/dataplane_helpers.go [hunk 4]
```diff
@@ -471,6 +467,10 @@ type TagSelectorRank struct {
 	WildcardMatches int
 }
 
+func (r TagSelectorRank) Zero() bool {
```
Comment (lobkovilya) [inline api/mesh/v1alpha1/dataplane_helpers.go:470]: Is it used? do we need it?

File: pkg/core/resources/apis/mesh/virtual_outbound_helpers.go [hunk 5]
```diff
@@ -0,0 +1,77 @@
...
+	"strconv"
+	"strings"
+	"text/template"
+
+	"github.com/asaskevich/govalidator"
+
+	mesh_proto "github.com/kumahq/kuma/api/mesh/v1alpha1"
+)
+
+func tagKeyOrName(parameter *mesh_proto.VirtualOutbound_Conf_TemplateParameter) string {
```
Comment (lobkovilya) [inline pkg/core/resources/apis/mesh/virtual_outbound_helpers.go:14]: If I understand it correctly we require to add a `parameter` entry even if a tag has the same key as a parameter in gotemplate?
So if we have template:
```yaml
host: {{ .foo }}.{{ .bar }}.{{ .baz }}
```
and we have DPP with inbound tags:
```yaml
tags:
  foo: value1
  bar: value2
  baz: value3
```
Anyway I have to provide parameters section:
```yaml
host: {{ .foo }}.{{ .bar }}.{{ .baz }}
parameters:
  - name: foo
  - name: bar
  - name: baz
```
Is it possible to resolve a template without explicit definitions of the parameters if no mapping required?

**Pull request merged**